### ISSUE TITLE
CRNN 语料

### ISSUE THREAD

**Issue by xiaomaxiao** [chineseocr/chinese-ocr#1]
crnn训练语料?可否共享下。

**Comment by wenlihaoyu** [comment]
语料是通过爬虫，爬取各新闻网站的新闻数据。代码如下：
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
# Created on 2017-10-20 22:02:59
# Project: news

from pyspider.libs.base_handler import *
from bs4 import BeautifulSoup
import uuid

class Handler(BaseHandler):
    crawl_config = {
    }

    @every(minutes=24 * 60)
    def on_start(self):
        self.crawl('http://news.qq.com/', callback=self.index_page)

    @config(age=10 * 24 * 60 * 60)
    def index_page(self, response):
        for each in response.doc('a[href^="http"]').items():
            self.crawl(each.attr.href, callback=self.detail_page)

    @config(priority=2)
    def detail_page(self, response):

        for each in response.doc('a[href^="http"]').items():
            self.crawl(each.attr.href, callback=self.next_detail_page)
            
    def next_detail_page(self, response):
        regex(response)
        
def regex(response):
        text = BeautifulSoup(response.text)
        ptext = text.find_all('p')
        chars = ''
        for p in ptext:
            chars+=p.text+'\n'
        if chars!='':
            f = open('/tmp/txt/txt_{}.txt'.format(uuid.uuid1().__str__()),'w')
            f.write(chars.encode('utf-8'))

**Comment by xiaomaxiao** [comment]
@wenlihaoyu  字频平衡如何控制？

**Comment by wenlihaoyu** [comment]
每取一次语料，都会统计每个字符频率，如果发现该段字符串的某个字符的频率大于平均频率的五倍，那么就舍弃，这样可以达到近似平衡，后面发现很多字符还是偏少（不常用的字符），就选择了一些古文构造了一部分数据，这样就解决双尾分布问题。

**Comment by wenlihaoyu** [comment]
用pytorch.crnn训练的模型更平稳一些（crnn目录下），keras太耗资源，当loss<0.2就中止了。代码里面两种模型都可以选择。

**Comment by fendaq** [comment]
可否共享一些训练集数据样例上来，谢谢

**Comment by wenlihaoyu** [comment]
后面有时间的时候，共享出来

**Comment by xiaomaxiao** [comment]
@wenlihaoyu  keras 需要把最后一个BN放在conv后，就好了。

**Comment by YCG09** [comment]
@xiaomaxiao 请问你说的是什么意思，现在最后一个BN不就在conv后吗？

    m = Conv2D(512,kernel_size=(3,3),activation='relu',padding='same',name='conv5')(m)
    m = BatchNormalization(axis=1)(m)
    m = Conv2D(512,kernel_size=(3,3),activation='relu',padding='same',name='conv6')(m)
    m = BatchNormalization(axis=1)(m)
    m = ZeroPadding2D(padding=(0,1))(m)
    m = MaxPooling2D(pool_size=(2,2),strides=(2,1),padding='valid',name='pool4')(m)
    m = Conv2D(512,kernel_size=(2,2),activation='relu',padding='valid',name='conv7')(m)

**Comment by YCG09** [comment]
@xiaomaxiao 我理解你是说改成这样，不知道对不对

m = Conv2D(512,kernel_size=(3,3),activation='relu',padding='same',name='conv5')(m)
m = BatchNormalization(axis=1)(m)
m = Conv2D(512,kernel_size=(3,3),activation='relu',padding='same',name='conv6')(m)
m = ZeroPadding2D(padding=(0,1))(m)
m = MaxPooling2D(pool_size=(2,2),strides=(2,1),padding='valid',name='pool4')(m)
m = Conv2D(512,kernel_size=(2,2),activation='relu',padding='valid',name='conv7')(m)
m = BatchNormalization(axis=1)(m)

**Comment by xiaomaxiao** [comment]
@YCG09 https://github.com/xiaomaxiao/keras_ocr/blob/master/CRNN/crnn.ipynb

**Comment by YCG09** [comment]
@xiaomaxiao 谢谢，但为什么axis=3呢，不应该是1吗

m = BatchNormalization(axis=3)(m)

**Comment by daiw10** [comment]
@YCG09  我理解tensorflow backend的话应该是3，channel层做bn，作者的可能是channel first

**Comment by YCG09** [comment]
@daiw10 恩，改后训练变快了很多
